### ISSUE TITLE
Improve relations to play nicer with server APIs (e.g. Retrofit, GSON, Jackson, Moshi, ...)

### ISSUE THREAD

**Issue by greenrobot** [greenrobot/ObjectBox#104]
If you get object data from another source (e.g. server) you can make those data objects also ObjectBox entities. One thing has to be considered: relations on ObjectBox currently rely on `ToOne` and `ToMany` types, which 3rd party libraries are not aware of.

A quick work around would probably be to walk through the object graph and replace `List` objects with `ToMany` objects.

However, we could also look into supporting plain `java.util.List`s in some way. A `put` would have to detect the type and do some additional syncing to figure out what to do. Or maybe just plainly `put` all objects in the list?

**Comment by FabianTerhorst** [comment]
That would fix #95 for me.

**Comment by greenrobot** [comment]
Not 100% certain how to implement it. Somewhat hesitant to enable putting entire lists as this may "encourage" it, which just might be wrong in other scenarios. Maybe having a new marker annotation to enable it?

**Comment by FabianTerhorst** [comment]
An annotation for automatic inserting lists would be great.

**Comment by YBthebest** [comment]
Is there any news on this? Currently running into this exact same issue (of toOne's not being recognised by a 3rd party), and would love to keep on using ObjectBox without having to build costly wrappers around everything...

**Comment by Vincent-Carrier** [comment]
Seconded. Meanwhile, by curiosity, what is the fix you guys are using?
Something like `order.forEach { order -> order.customer.target = order }` ?

**Comment by YBthebest** [comment]
Yep, with .setTarget(order). Seems like the right way around this for the moment.

**Comment by barykaed** [comment]
Yes please. Would like to use an ObjectBox "entity" as a model of Retrofit responses too. Using Moshi for the json adapter. Please suggest a way to use these together.

**Comment by ecramer** [comment]
Any ETA on this? Looking to move from Realm to ObjectBox but this issue is making it difficult.

**Comment by ArthurSav** [comment]
I concur in regards to having an annotation based solution `@ToOne` `@ToMany`
POJOS or in this case entities should not be closely tied to a database.

Having something like:

`@ToOne
  List<Stuff> stuff
`

Would solve many of the issue here

**Comment by indrakumarprajapat** [comment]
I have a class with ToMany object, and using retrofit as rest client. I am getting this error while loading data,

**FATAL ERROR field com.company.entity.Receipt.receipt_items has type io.objectbox.relation.ToMany, got java.util.ArrayList**
 
This is my class structure.

```
@Entity
public class Receipt implements Serializable {

    public ToMany<ReceiptItem> receipt_items;

    @Id
    public long id;
    public long receipt_no;
}
```
Issue is arraylist not mapping to ToMany object.

**Comment by greenrobot-team** [comment]
@indrakumarprajapat Is this a JSON parsing error? For example with GSON you might have to create a [type adapter](url) for `ToMany`. -ut

**Comment by indrakumarprajapat** [comment]
okay, I got it, but can you help me, how can I do that. i have little bit idea. but not sure how to do it exacly.

**Comment by greenrobot-team** [comment]
@indrakumarprajapat Sorry, not hear. Please look in the GSON documentation or ask for example on Stack Overflow for help. -ut

**Comment by indrakumarprajapat** [comment]
Sorry for late reply, the issue is fixed by type adapter.  thanks.

**Comment by ferdavs** [comment]
I don't like this solution but it works.
you have to implement JsonDeserializer.

```java
    public  class MyDeserializer<T extends FromJson> implements JsonDeserializer<T> {
        @Override
        public T deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
            try {
                T var = ((Class<T>) typeOfT).newInstance();
                var.fromJson(json.getAsJsonObject());
                return var;
            } catch (InstantiationException e) {
                e.printStackTrace();
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            } catch (Throwable e) {
                e.printStackTrace();
            }

            return null;
        }
    }
```
create Gson object to use with Retrofit 
```java
 Gson gson= new GsonBuilder()
                .registerTypeHierarchyAdapter(FromJson.class, new MyDeserializer())
                .create();

Retrofit retrofit = new Retrofit.Builder()
                .client(client)
                .baseUrl(baseUrl)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();
```

FromJson interface and some example.
```java
public interface FromJson {
    void fromJson(JsonObject jo) throws JSONException;
}

@Entity
public class Status implements FromJson  {
    public long id;
    public int code;
    public String message;

    public void fromJson(JsonObject jo) throws JSONException {
            this.code = jo.get("code").getAsInt();
            this.message = jo.get("message").getAsString();
    }
}

@Entity
public class Login implements FromJson  {
    public long id;
    public ToOne<User> user;
    public ToOne<Status> status;

    @Override
    public void fromJson(JsonObject jo) throws JSONException {
            Status s = new Status();
            s.fromJson(jo.get("status").getAsJsonObject());
            this.status.setTarget(s);
        
            User u = new User();
            u.fromJson(jo.get("user").getAsJsonObject());
            this.user.setTarget(u);
    }
}

```

Any example using TypeAdapter? example provided by @greenrobot-team is dead.

**Comment by Haskell2015** [comment]
@indrakumarprajapat How does the TypeAdapter work? Could U show me a simple demo ? Thks :)

**Comment by markwhitaker** [comment]
I'm not sure any of the comments here directly reflect the case of simple nested objects, of the type that crop up all the time in JSON data feeds, e.g.:

```json
{
   "name": "Acme Inc.",
   "location": {
      "longitude": 1.23456,
      "latitude": 6.54321
   }
}
```

Which might deserialize to...

```java
public class Location
{
   private double longitude;
   private double latitude;
   // getters, setters etc.
}

@Entity
public class Business
{
   private String name;
   // HOW TO PERSIST THIS??
   private Location location;
}
```

These are a problem too, but we don't usually need an entity relationship here. Something more like Room's [`@Embedded`](https://developer.android.com/reference/android/arch/persistence/room/Embedded.html) annotation would be sufficient.

(I think this is more accurately covered by https://github.com/objectbox/objectbox-java/issues/217 which is marked as a duplicate of this issue.)

**Comment by ferdavs** [comment]
[objectbox+Retrofit](https://github.com/ManueGE/Boxfit)

**Comment by lgengsy** [comment]
ToMany and  ToOne is so bad !!!

**Comment by lgengsy** [comment]
I gave up because of this problem, and I don't think this data structure design is reasonable enough.

**Comment by greenrobot** [comment]
@lgengsy Could you provide a little bit info for us, so we can fix it?

**Comment by uiiang** [comment]
@indrakumarprajapat How does the TypeAdapter work? Could U show me a simple demo ? Thks :)

**Comment by greenrobot** [comment]
Quick poll: Do you mostly need this for `List`? Or is `ToOne` a hard requirement?

**Comment by lgengsy** [comment]
At first, I thought this project could help me deal with data caching problems quickly. I tried it. When I met toMany and toOne, I found the project was getting more and more complicated. It took me a day or two to adjust my data structure. Finally, I found it too hard to change back.

**Comment by Riabkov** [comment]
Any new solutions or workarounds? Or better look for another ORM? (Room?)

**Comment by hetang** [comment]
Any solutions for ToOne or ToMany to work with Retrofit?

**Comment by MhmDSmdi** [comment]
Is there any solution for deserialize ToOne Object Using Gson Converter ?
@greenrobot 
@greenrobot-team

**Comment by suleymanccelik** [comment]
Is there any update or any solution for working with Gson?

**Comment by wb1992321** [comment]
@greenrobot 
@greenrobot-team 
Can you summarize the specific implementation? It is best to add to the document

**Comment by Xset-s** [comment]
I have the same issue. Please, implement annotation for ToOne.

com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `io.objectbox.relation.ToOne` (no Creators, like default construct, exist): cannot deserialize from Object value (no delegate- or property-based Creator)

**Comment by Xset-s** [comment]
Any news??

**Comment by MhmDSmdi** [comment]
I can't fix the problem of Objectbox with GSON for deserialize ToOne Object but I'm using Moshi now and fix it by some trick.
1.you must create same model class but Instead of ToOne, declare main Class.(ObjectResponse)
2.create a JsonAdapter class and use @FromJson , @ToJson method for convert Json to your object
3.configure Moshi Object (add JsonAdapter class to Moshi)
for more information see this: []( https://github.com/square/moshi#another-example)

**Comment by VenkateshPrasadSV** [comment]
@greenrobot Is there any update for this issue? Manually parsing JSON is becoming difficult. I have used a workaround for saving the list but assume if we have a list inside another list, it becomes more complex to solve with the workaround.

**Comment by racofix** [comment]
@greenrobot @greenrobot-team How to solve the issue??

**Comment by greenrobot-team** [comment]
Recently had to work with ObjectBox and Gson. If you don't want to add temporary fields the only other solution I could come up with is to write a custom deserializer for each entity.
```java
@Entity class User {
    private ToOne<Status> status;
    private ToMany<Address> addresses;
    // TODO getters/setters
}

class UserDeserializer implements JsonDeserializer<User> {
    @Override
    public User deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
            throws JsonParseException {
        JsonObject userJson = json.getAsJsonObject();
        User user = new User(); // ToOne and ToMany initialized here through code injected by ObjectBox plugin

        // ToOne
        JsonObject statusJson = userJson.getAsJsonObject("status");
        Status status = context.deserialize(statusJson, Status.class);
        user.getStatus().setTarget(status);

        // ToMany
        JsonArray addressesJson = userJson.getAsJsonArray("addresses");
        Type addressListType = new TypeToken<ArrayList<Address>>(){}.getType();
        ArrayList<Address> addresses = context.deserialize(addressesJson, addressListType);
        user.getAddresses().addAll(addresses);

        // TODO parse other properties

        return user;
    }
}
```
-Uwe

**Comment by kibotu** [comment]
are there any plans on having auto generated type adapters or any reducing boilerplate implementations in the future?

**Comment by ArthurSav** [comment]
One drop of oil can ruin a tank full of water. 

Revisited ObjectBox after 2 years in hopes of seeing some improvements but i see this issue (among others) still persists.

I'm afraid issues like this is a deal breaker for many mobile developers that care about quality code. It's not just the boilerplate code to deal with **ToOne** **ToMany**, it's also that my app has to be aware of unnecessary business logic of a 3rd party lib in many places.

Maybe in another 2 years :)

**Comment by francispinsoro** [comment]
I develop this solution, based on this post.
http://www.jhr8.com/p/574396eb1412?utm_campaign=maleskine&utm_content=note&utm_medium=seo_notes&utm_source=recommendation

I assing to gson a deserializer of a invented class 'JsonBox', and all entities extends from that class, and i change the statement 'ToMany', through the interface 'List'.

This is the deserializer:

`public class JsonBoxDeserializer<T extends JsonBox> implements JsonDeserializer<T> {

    private Gson gsonBox;

    public JsonBoxDeserializer(){
        gsonBox = MyGson.getGsonBox();
    }

    @Override
    public T deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        try {
            T var = gsonBox.fromJson(json, typeOfT);
            for (Map.Entry<String, JsonElement> entry : json.getAsJsonObject().entrySet()) {
                try {
                    Field field = var.getClass().getField(entry.getKey());
                    if (field.getType() == List.class) {
                        filling((List) field.get(var), ((Class) (((ParameterizedType) field.getGenericType()).getActualTypeArguments()[0])), entry.getValue(), context);
                    } else if (field.getType() == ToOne.class) {
                        filling((ToOne<? extends JsonBox>) field.get(var), ((Class) (((ParameterizedType) field.getGenericType()).getActualTypeArguments()[0])), entry.getValue(), context);
                    }

                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
            return var;
        } catch (Throwable e) {
            e.printStackTrace();
        }
        return null;
    }

    protected <T> void filling(List<T> toMany, Class<T> type, JsonElement jsonElement, JsonDeserializationContext context) {
        if (toMany == null)
            return;
        toMany.clear();
        if (jsonElement.isJsonArray()) {
            JsonArray array = jsonElement.getAsJsonArray();
            for (int i = 0; i < array.size(); i++) {
                JsonElement element = array.get(i);
                T item = context.deserialize(element, type);
                toMany.add(item);
            }
        }
    }

    protected <T extends JsonBox> void filling(ToOne<T> toOne, Type typeOfT, JsonElement jsonElement, JsonDeserializationContext context) {
        if (toOne == null)
            return;
        T one = context.deserialize(jsonElement,typeOfT);
        toOne.setTarget(one);

    }

}`

I instance gson excluding List and ToOne.

` public static Gson getGsonBox(){
        return new GsonBuilder().setExclusionStrategies(new ExclusionStrategy() {
            @Override
            public boolean shouldSkipField(FieldAttributes f) {
                return false;
            }

            @Override
            public boolean shouldSkipClass(Class<?> clazz) {
                return clazz == List.class || clazz == ToOne.class;
            }
        }).create();
    }`

And the first deserializer to call:

`new GsonBuilder()
                .registerTypeHierarchyAdapter(JsonBox.class, new JsonBoxDeserializer<>())
                .create();`

**Comment by damonbakker** [comment]
This was a core part of my work with greendao. 

I did a lot of syncing with retrofit on a lot of endpoints.
Spend a good amount of time refactoring to allow for objectbox with daocompat.

Would really love to see some compatibility here.
I'll try to implement some of the suggestions above, potentially reverting to greendao if it doesn't work out nicely.

**Comment by FlodCoding** [comment]
> http://www.jhr8.com/p/574396eb1412?utm_campaign=maleskine&utm_content=note&utm_medium=seo_notes&utm_source=recommendation



> I develop this solution, based on this post.
> http://www.jhr8.com/p/574396eb1412?utm_campaign=maleskine&utm_content=note&utm_medium=seo_notes&utm_source=recommendation
> 
> I assing to gson a deserializer of a invented class 'JsonBox', and all entities extends from that class, and i change the statement 'ToMany', through the interface 'List'.
> 
> This is the deserializer:
> 
> ```
> public class JsonBoxDeserializer<T extends JsonBox> implements JsonDeserializer<T> {
> 
>     private Gson gsonBox;
> 
>     public JsonBoxDeserializer(){
>         gsonBox = MyGson.getGsonBox();
>     }
> 
>     @Override
>     public T deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
>         try {
>             T var = gsonBox.fromJson(json, typeOfT);
>             for (Map.Entry<String, JsonElement> entry : json.getAsJsonObject().entrySet()) {
>                 try {
>                     Field field = var.getClass().getDeclaredField(entry.getKey());
>                     field.setAccessible(true);
>                     if (field.getType() == List.class) {
>                         filling((List) field.get(var), ((Class) (((ParameterizedType) field.getGenericType()).getActualTypeArguments()[0])), entry.getValue(), context);
>                     } else if (field.getType() == ToOne.class) {
>                         filling((ToOne<? extends JsonBox>) field.get(var), ((Class) (((ParameterizedType) field.getGenericType()).getActualTypeArguments()[0])), entry.getValue(), context);
>                     }
> 
>                 } catch (Exception e) {
>                     e.printStackTrace();
>                 }
>             }
>             return var;
>         } catch (Throwable e) {
>             e.printStackTrace();
>         }
>         return null;
>     }
> 
>     protected <T> void filling(List<T> toMany, Class<T> type, JsonElement jsonElement, JsonDeserializationContext context) {
>         if (toMany == null)
>             return;
>         toMany.clear();
>         if (jsonElement.isJsonArray()) {
>             JsonArray array = jsonElement.getAsJsonArray();
>             for (int i = 0; i < array.size(); i++) {
>                 JsonElement element = array.get(i);
>                 T item = context.deserialize(element, type);
>                 toMany.add(item);
>             }
>         }
>     }
> 
>     protected <T extends JsonBox> void filling(ToOne<T> toOne, Type typeOfT, JsonElement jsonElement, JsonDeserializationContext context) {
>         if (toOne == null)
>             return;
>         T one = context.deserialize(jsonElement,typeOfT);
>         toOne.setTarget(one);
> 
>     }
> 
> }
> ```
> 
> I instance gson excluding List and ToOne.
> 
> ```
> public static Gson getGsonBox(){
>         return new GsonBuilder().setExclusionStrategies(new ExclusionStrategy() {
>             @Override
>             public boolean shouldSkipField(FieldAttributes f) {
>                 return false;
>             }
> 
>             @Override
>             public boolean shouldSkipClass(Class<?> clazz) {
>                 return clazz == List.class || clazz == ToOne.class;
>             }
>         }).create();
>     }
> ```
> 
> And the first deserializer to call:
> 
> `new GsonBuilder() .registerTypeHierarchyAdapter(JsonBox.class, new JsonBoxDeserializer<>()) .create();`

这个链接真牛啊，看起来不像简书啊（https://www.jianshu.com/）

**Comment by ajans** [comment]
I wanted to give my point of view on how I handle this scenario:

I was pondering how to do this and what would be considered as a best practice.
This article gave some insightful pointers on why it is not a good idea to use the Retrofit-response-model as the model for ObjectBox-operations:
https://proandroiddev.com/the-real-repository-pattern-in-android-efba8662b754

TLDR:

- Models used for remote communication (Retrofit-models aka DTOs (data transfer objects)) should not be used as domain-model entities used by an app, as changes to the DTO would affect the whole codebase for the app
- Therefore, there should be a mapping-process between DTOs and domain-model entities and a repository-class should always return domain-model entities to the app
- This way, if anything changes on the backend-side, only the mapping in the respective repository-class has to be adapted

With this way of thinking, I was able to mitigate this issue in a clean way without having to resort to custom deserializers or other hacky solutions.
So, my BaseRepository-code to refresh a list of data from a remote API and save it to ObjectBox basically looks something like this:

```
/**
 * Type parameters:<br/>
 * DTO: domain transfer object (remoteapi-gson-models)br/>
 * DM: domain model (app-models)<br/>
 */
abstract class BaseRepository<DTO, DM : DomainModel>(
    val service: RemoteApi,
    private val localDataSource: LocalDataSource<DM>
) {
    // ...
    suspend fun refreshList() {
        wrapEspressoIdlingResource {
            val dtoData: Result<List<DTO>?> = fetchList()
            if (dtoData is Result.Success) {
                val mappedData = mapList(dtoData.data)
                saveToDatabase(mappedData)
            } else if (dtoData is Result.Error) {
                throw dtoData.exception
            }
        }
    }

    fun mapList(dtoData: List<DTO>?): List<DM>? {
        logTrace()
        if (dtoData == null || dtoData.isEmpty())
            return null

        val result: ArrayList<DM> = ArrayList()
        for ((index, sourceItem) in dtoData.withIndex()) {
            val mappedModel = map(index, sourceItem)
            result.add(mappedModel)
        }

        return result
    }

    abstract fun map(index: Int, dtoItem: DTO): DM
```

`refreshList` is used from a ViewModel-LiveData.

First step of `refreshList` is to get the DTO-data from the remote-API via `fetchList`.

Second step is to map this data with the `mapList` and `map` functions.
The concrete implemention of the map-function instantiates the domain-model-entities, sets their properties and adds related mapped sub-entities and puts them into the `ToMany`-fields.

After `mapList`, the mapped data (in your domain-model entities) then gets persisted into the ObjectBox-database in the `saveToDatabase`-method.

There is a very similar implementation of this steps for single DTO-item-responses from the remote-API.

To use this code, you derive a Repository-class from BaseRepository for a pair of DTO and domainmodel-entity that you want to persist in your objectbox from this. In the concrete Repository, you implement the map-function, the remote-API call (called from fetchList) and define the generic type of the LocalDataSource-interface in the constructor. The concrete LocalDataSource-implementation class for the used ObjectBox-entity knows the specific Box or Boxes for used entities and uses those boxes to get or put the entity-instances or provide ObjectBoxLiveData-instances with the specific queries.

If you read the article above and its recommendations, you might think this is a bit of cheating. That is true, if you would follow the recommendations consequently, you would not use ObjectBox-annotations and ToMany in your domain-model entities at all (keep them pure of any persistence and communication framework stuff), but I found it works quite well in my case as a tradeoff between pragmatism and purity as I didn't want to introduce another model-layer and its mapping at the time.

The one thing I had a bit of trouble with, was unit-testing the ViewModels and the repository-classes' mapping with ObjectBox running on desktop, as in my multi-module project, the ToMany-property would not get initialised and failed the tests.

I worked around that by abstracting ObjectBox away behind a LocalDataSource-interface (injected into the Repository-constructor) , thus faking it for the unit-tests and doing the "Initialization magic" trick mentioned in the objectbox-docs.

So, my TLDR recommendation is:

Don't use your Gson- or Moshi-models as ObjectBox-models.

At least separate those two data-source models by mapping one onto the other in your repository-classes.
This way, you can use ObjectBox the way you like and can even separate your remote-API-client out of your app-project into its own project that you don't have to build all the time but only if something changes on your backend.

P.S.: if you wonder why the `map`-function takes an incremented index as a first parameter, this is being used in a `@BaseEntity` `order`-property to preserve the order of DTO-items coming from the remote-API.
So if you have something like top-items with an arbitrary serverside-defined order to them, you don't have to duplicate that ordering-business-logic in your app in the ObjectBox-query. You can simply use the `order`-property.
You could also use ObjectBox-generated IDs, but I found this to be quite helpful with assignable IDs in my ObjectBox-models.

**Comment by Mustafa-Altameemi** [comment]
At the final stage ToMany<> just messed up on all my project.

**Comment by Ditscheridou** [comment]
Any updates here? Im currently looking into Objectbox + mapstruct, this ToMany/ToOne messes up my whole mapping process.

**Comment by Ditscheridou** [comment]
@greenrobot-team ?

**Comment by mecoFarid** [comment]
@ArthurSav 

> Maybe in another 2 years :)

Well, here is your 2 years

**Comment by mecoFarid** [comment]
> Recently had to work with ObjectBox and Gson. If you don't want to add temporary fields the only other solution I could come up with is to write a custom deserializer for each entity.
> 
> ```java
> @Entity class User {
>     private ToOne<Status> status;
>     private ToMany<Address> addresses;
>     // TODO getters/setters
> }
> 
> class UserDeserializer implements JsonDeserializer<User> {
>     @Override
>     public User deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
>             throws JsonParseException {
>         JsonObject userJson = json.getAsJsonObject();
>         User user = new User(); // ToOne and ToMany initialized here through code injected by ObjectBox plugin
> 
>         // ToOne
>         JsonObject statusJson = userJson.getAsJsonObject("status");
>         Status status = context.deserialize(statusJson, Status.class);
>         user.getStatus().setTarget(status);
> 
>         // ToMany
>         JsonArray addressesJson = userJson.getAsJsonArray("addresses");
>         Type addressListType = new TypeToken<ArrayList<Address>>(){}.getType();
>         ArrayList<Address> addresses = context.deserialize(addressesJson, addressListType);
>         user.getAddresses().addAll(addresses);
> 
>         // TODO parse other properties
> 
>         return user;
>     }
> }
> ```
> 
> -Uwe

We have more than 800 models tied with `ToOne/ToMany`relations, would be fun to implement what you recommend. One more thing; we must not forget to do the same for future relations or else the whole project will be doomed

**Comment by Mustafa-Altameemi** [comment]
Well, as I have gone deep through the library source code, so (I think) the only solution for @greenrobot-team  is to start building a new DB from scratch. It's very disappointing to see an experienced developers team fall into such a mistake. 

 Anyway thanks for your efforts.

**Comment by ajans** [comment]
> Well, as I have gone deep through the library source code, so (I think) the only solution for @greenrobot-team is to start building a new DB from scratch. It's very disappointing to see an experienced developers team fall into such a mistake.
> 
> Anyway @greenrobot-team thanks for your efforts.

As a long-time user of ObjectBox (currently working on the 3rd app with it, 2 of them have professional context), I am very curious to hear from you specific arguments based on specific ObjectBox-code to support your bold statement.

In my opinion it is a mistake to judge the whole ObjectBox-ecosystem based on a client-library for a specific platform (even if it is the primary and most mature one).

Yes, the handling of ToOne and ToMany-properties does not play very nice with e.g. Retrofit directly, but as I pointed out in my comment above, I think it is a fundamental design-flaw in one's app-code to not separate a Retrofit-POJO from an app-internal domain-model and its related ObjectBox-entity for the reasons stated in my earlier comment and the article I quoted.

**Comment by ArthurSav** [comment]
Forgot to unsubscribe from this. Here's my 2cents:

After many years of coding i realise why plain old sqlite is still relevant. The headaches caused by 3rd party libs are usually not worth the switch but that is obvious only after a lot of hard work has gone into a project.

**Comment by AfricanDev** [comment]
> ```java
> UserDeserializer 
> ```

can you provide sample on how to use UserDeserializer @greenrobot

**Comment by AfricanDev** [comment]
tried to add @JsonSerialize(using = UserDeserializer.class) but this does not work

**Comment by greenrobot-team** [comment]
@AfricanDev Can you please stop spamming on issues! It's enough to post once on one issue.

Look at the Gson user guide on how to de/serialize custom types. https://github.com/google/gson/blob/master/UserGuide.md

**Comment by richard457** [comment]
I did not see how this issue is so old given how this is critical my guess is that there is a real serious issue down in the codebase, I would like to know if there is any progress so far, this would help anyone in the future in decision making.

**Comment by greenrobot** [comment]
@ All: First, I'm really sorry that this hasn't been worked on in such a long time. Maybe we can find some minimal pragmatic solution to at least improve the situation? I've a couple of questions for you to find out...

Would you say that there's already value in offering an alternative to `ToMany` (without tackling `ToOne`) as a first step? 
That's be using `List` and when reading from the DB, it would be an instance of `ToMany`. When putting, we'd have to check the type and if it finds anything else than a `ToMany`, I think it makes sense to make this configurable.

E.g. `@Many(onPut = ManyStrategy.Merge)` or `@Many(onPut = ManyStrategy.Add)`, `@Many(onPut = ManyStrategy.Manual)`

"Merge" would be obviously the most complex and expensive solution, "add" would be add-only (may make sense if it's only to bootstrap data), and "manual" would do nothing, but leave the handling to you.

And finally: would it be enough to start with only `@Many(onPut = ManyStrategy.Manual)`? The less initial work, the more likely we can squeeze this in.

**Comment by richard457** [comment]
**Would you say that there's already value in offering an alternative to ToMany (without tackling ToOne) as a first step?** yes better than having none, I support anything to start with then we can experiment as we go, I prefer MVP and continuous improvement so yes I support it, Also I like confurable options since it give more room for improvement. I am not aware of the codebase so I can decide which is better than other but for now I would focus less on merge as you mentioned.

**Comment by Ditscheridou** [comment]
I would suggest to go with an iterative approach from simple to complex here, to improve the situation as fast as possible and implementing the ideal solution step by step. So starting with a Manual Type would be fine at first. The ideal thing would be a cascade merge, but i admit thats costly and maybe currently a little bit too much to ask for.

**Comment by ajans** [comment]
An annotation like @Many would be very appreciated, even for me with my mapping-approach described above.
As this annotation would remove the initialization-magic boilerplate in some entities.

@greenrobot I guess this approach would put most of the development work into the objectbox-java codegenerator part (the gradle-plugin)?
Would you be open to opensource that part of the objectbox-project to enable pull requests from the community?

I would love to take a shot at it, since it would be a good opportunity to dedicate my next learningday-friday this week to dig myself into how to enhance a gradle-plugin / code generator.
I also would be open to do pair-programming on this.

**Comment by greenrobot** [comment]
> would it be enough to start with only @Many(onPut = ManyStrategy.Manual)? 

Looking into this a little closer we found, that ObjectBox already behaves like proposed. :laughing:  E.g. you can declare a `List`, which will be ignored if it's not an instance of `ToMany`, which should be the case if the object is coming from e.g. a JSON lib. Please verify if that works for your use cases too.

Sound we should like update [relation docs](https://docs.objectbox.io/relations) to mention this more prominently.

We're still thinking about `@Many(onPut = ManyStrategy.Manual)` to enable this behavior explicitly and throw otherwise to ensure this does not happen by accident.

**Comment by ajans** [comment]
@greenrobot I don't yet understand how this would help.

If ObjectBox just ignores a `List`-property, that is not an instance of `ToMany`, wouldn't that mean I as an app-developer would have to:

- do the intialization-magic (instantiating the `ToMany`-instance and assigning it to the `List`-property) manually when instantiating the POJO and
- call ToMany.addAll with the `ArrayList`-instance every time just before I want to put the deserialised JSON-list into ObjectBox?

I just tried to change the `ToMany`-property to a `List`-property in my Kotlin-code (1.5.20) and the compiler is complaining about not being able to find a matching `addAll`-invariant:

Entity with `List`/`ToMany`-property `hits`:

>     @Entity
>     data class TopDeals(
>         @Id var id: Long = 0,
>         var totalHits: Int? = null,
>         var totalDealsApproximateCount: String? = null,
>     ) : BaseModel(), DomainModel {
>         // objectbox initialization magic does not work in TopDealsViewModelTest
>         // lateinit var hits: ToMany<TopDealProduct>
>         var hits: List<TopDealProduct> = ToMany(this, TopDeals_.hits)
>     }

Usage:

>         private fun mapTopDealProducts(dtoItem: TopDealsDTO, topDeals: TopDeals) {
>             dtoItem.hits?.map { dtoProduct ->
>                 TopDealProduct(
>                     productId = dtoProduct.productId,
>                     dealDiscount = dtoProduct.deal?.discount,
>                     dealDiscountAbsolute = dtoProduct.deal?.discountAbsolute,
>                     dealDiscountHotMarker = dtoProduct.deal?.discountHotMarker ?: false,
>                     dealDiscountTopMarker = dtoProduct.deal?.discountTopMarker ?: false,
>                     imageUrl = dtoProduct.image ?: URL_NOIMAGE,
>                     minPrice = dtoProduct.minPrice,
>                     minTotalPrice = dtoProduct.minTotalPrice,
>                     relevance = dtoProduct.relevance
>                 ).apply {
>                     mapSoluteEntity(dtoProduct, this)
>                 }
>             }?.let { topDealProducts -> topDeals.hits.addAll(topDealProducts) }
>         }

Error:

> 
> Unresolved reference. None of the following candidates is applicable because of receiver type mismatch: 
> public fun <T> MutableCollection<in TypeVariable(T)>.addAll(elements: Array<out TypeVariable(T)>): Boolean defined in kotlin.collections
> public fun <T> MutableCollection<in TypeVariable(T)>.addAll(elements: Iterable<TypeVariable(T)>): Boolean defined in kotlin.collections
> public fun <T> MutableCollection<in TypeVariable(T)>.addAll(elements: Sequence<TypeVariable(T)>): Boolean defined in kotlin.collections

Tested with ObjectBox 2.9.1

**Comment by greenrobot-team** [comment]
@ajans Maybe to clarify: it is currently possible to replace the `List` (or `MutableList` in Kotlin) instance, which is a `ToMany` when creating an object, with something like `ArrayList`. Replacing the instance is what a typical JSON deserializer does, no?

If the object is then put, the code will silently skip the `List`/to-many property. It is then up to the developer to create the relations in whatever way.

The current suggestion is to not make the put silently skip, but error, and require adding the `@Many` annotation to again silently skip the property on put.

**Comment by greenrobot-team** [comment]
Also added a section to the to-many docs about using `List`: https://docs.objectbox.io/relations#using-the-list-interface-for-to-many

**Comment by AfricanDev** [comment]
in the doc we have this example ```// Teacher.java
@Entity
public class Teacher{    
    @Id public long id;    
}

// Student.java
@Entity
public class Student{    
    @Id public long id;    
    public List<Teacher> teachers = new ToMany<>(this, Student_.teachers);   
}``` where  Student_.teachers comes from?

**Comment by greenrobot-team** [comment]
@AfricanDev Can you explain how this is related to this issue? These meta-information classes with the underscore (`Student_`) are generated by the ObjectBox plugin on building the project for each entity. They are mentioned on the Queries docs as they are typically used in queries: https://docs.objectbox.io/queries#querybuilder

**Comment by greenrobot-team** [comment]
> So if I understand you correctly, you are saying, as of objectbox-version>=3.3.1, during app-runtime, as long as nothing overwrites a List-property pointing to another objectbox-entity, objectbox-initialization-magic picks this property up and uses it like a `ToMany`-instance as usual?

Yes.

But so we are on the same terms: we refer to "initialization magic" as the process where ToOne and ToMany are instantiated when constructing a new object (the magic being the ObjectBox byte code transformer adding the initialization code to your objects constructors).

The code that "picks this property up" is the Cursor classes. [When doing put they use a ToMany API to update the relation.](https://github.com/objectbox/objectbox-java/blob/d1883182f1b3687948e97c1e694ee67c7b09e645/objectbox-java/src/main/java/io/objectbox/Cursor.java#L330-L339) But if the relation property isn't a `ToMany` instance nothing happens:

https://github.com/objectbox/objectbox-java/blob/d1883182f1b3687948e97c1e694ee67c7b09e645/objectbox-java/src/main/java/io/objectbox/Cursor.java#L330-L339

So [this is fine,](https://docs.objectbox.io/relations#using-the-list-interface-for-to-many) as long as `teachers` remains a `ToMany` instance:

```kotlin
var teachers: MutableList<Teacher> = ToMany<>(this, Student_.teachers)
```

Actually I lied, it is possible to overwrite a `ToMany` instance when the relation is empty (as stored in the database). This is because the job of a `ToMany` is to keep track of changes, e.g. when adding or removing a related object. So when creating a new object or getting one that doesn't have related objects it's possible to replace the `ToMany` instance at any point.

E.g. technically this should be possible (again assuming it's a new object or for an existing object where `teachers` is empty):
```kotlin
teachers = listOf(teacher1, teacher2)
val toMany = ToMany<>(this, Student_.teachers)
toMany.addAll(teachers)
teachers = toMany
```
